### DISCUSSION TITLE
Reduce amount of writes to SD-card (to protect it from wearing out)

### DISCUSSION THREAD

**Discussion by d1vzero** [dtcooper/raspotify#567] [Ideas]
First of all, thanks a lot for the excellent work!

I am using raspotify since years on a raspberryPi. Recently on my pi3 my SD-card died because of too many write operations over the last years. On my new setup (same pi3 with rasberrypios 11, bullseye) I wanted to improve things and reduce the number of dics writes.
Using iotop, i figured out, that a large portion of the dics writes is caused by librespot. Hence, I carefully checked that audio caching is really disabled. However, the dics writes were caused by the currently played song. The song is not directly streamed, but temporarily "downloaded" to the path given in librespots environment variable TMPDIR and deleted directly afterwards. This causes a high IO on the SD-card.

To circumvent this behaviour the idea was to write the currently played song directly to a RAM-disc (tmpfs) instead of the SD-card. So here is what I did.

As mentioned, per default librespot writes the downloaded file to its temporary directory given in the Environment variable TMPDIR. So in raspotifys systemd service one simply has to point that variable to a path located in a tmpfs. To make the changes persistent and not being overwritten with the next raspotify update do not edit raspotifys systemd service configuration file directly, but create an override file using:

`sudo systemctl edit raspotify.service`

This generates an override file and automatically opens that file in nano. Simply add
```
[Service]
RuntimeDirectory=%N
Environment=TMPDIR=%t/%N
```
and save the file. The two lines define a runtime directory which is mounted as tmpfs (ramdisc) located at /run/raspotify and sets the Environment variable TMPDIR to exactly that location. Just save and issue the commands
```
sudo systemctl daemon-reload
sudo systemctl restart raspotify.service
```

From now on the file will only be written to RAM which protects your SD-card from write cycles when music is played back.

If you additionally want to control whether to write to SD-card or to RAM you can add the following line to your /etc/raspotify/conf
`TMPDIR=/tmp`

If uncommented raspotify writes to the SD-card directly (to its private temp directory at /tmp/systemd-private-[hash]-raspotify.service).
If commented it writes to a tmpfs located at /run/raspotify

This is what helped for me, maybe its worth thinking about to include it in the official release, which makes the systemd override file unnecessary.

**Comment by JasonLG1979** [comment]
I/O isn't really an issue if you get a decent SD card. I've had the same SD card in 4 different versions of Pi's going back for over 5 years and have re-imaged and abused them more than I care to admit. If disk I/O is such a big deal you should just set the Pi to boot read only in raspi-config.

My gut tells me that you're entering unnecessary optimization territory. 

An alternative if you have a reasonable sized SD-card that will reduce disk and network I/O and theoretically boost performance a tiny bit is to configure librespot to cache audio files to disk. I have it set up with a cache size of 2 GB (16 GB SD card) on my Pi Zero v2 and 16 GB on my home server (the server is not a Pi) You can fit a lot of 320kbps Vorbis' in 2 GB that later do not need to be re-downloaded.

That being said I'm open to a tmpfs as default so long as it doesn't break disk caching or otherwise negatively affect the operation of Raspotify in any way.

Put in a PR and I will review it.

**Comment by d1vzero** [comment]
Well, so far I also never had issues with any SD-card, but now it appeared very suddenly and slowed down things extremely. However, I have to admit that I used the SC-card for quite some years reformatting it several times. Of course this also depends on how occupied the card is. Since a lot of people even advise to put /var/log on a tmpfs I wanted to check the I/O in advance. During normal usage I found, that librespot by far caused the highest I/O for me. 
Maybe you're right and it 's over optimization, who knows exactly. On the other hand I haven't seen any drawbacks yet. One point I haven't carefully checked is, how very large songs or even long podcasts are handled during playback.

I will put in a PR and we will see.

**Reply by JasonLG1979** [reply]
I tested it out for a couple hours on my Pi Zero v2 it didn't seem to negatively effect it in any way.

> One point I haven't carefully checked is, how very large songs or even long podcasts are handled during playback.

I don't listen to podcasts ever so I don't know how long they can get? But according to [this audio file size calculator](https://www.colincrawley.com/audio-file-size-calculator/) the track would have to be over 2 hours long to even come close to exhausting the RAM on a Pi with 512MB.

I don't think I'll personally be using it but what we can do is make it so it defaults to using `TMPDIR=/tmp` in the config so users have the option if they'd like because I do see this as something some people (like yourself) would want.

**Comment by JasonLG1979** [comment]
Done:
https://github.com/dtcooper/raspotify/commit/68f45add8e9d94af3c01c026ddd2e5c951445ba6

**Reply by d1vzero** [reply]
Oh, that was a way faster than I was able to even react. Thanks anyhow. Best regards

**Reply by JasonLG1979** [reply]
Well it was a reasonable, in-scope request and something that other people might want that doesn't change the default behavior. The perfect kind of feature request.

You'd be surprised how often people file bug reports for things that are totally not bugs or are caused by user error and/or failure to actually read the README or wiki, or make feature requests that are super niche or completely out of scope or out of my control.

Basically it's a lot of "I can't get it to work the way I want and I didn't read any of the docs can you spoon feed me." or "I want Raspotify to make me a sandwich."

**Reply by d1vzero** [reply]
Sorry, I was off for some days.
Well, sounds funny how you give it a name, but I can entirely feel what you are saying.
By the way, that was my point, why I shared this idea. So far I took a lot from the FOSS community and that time it was perfect to give something back, even though it's just two or three tiny lines of "code".

So, thanks again and keep ongoing with your good work and maintenance.

**Reply by JasonLG1979** [reply]
Download buffering is what Gstreamer calls it and also does the same thing for remote files on occasion.

**Reply by d1vzero** [reply]
Well then let's name it accordingly.

**Comment by dowaichi** [comment]
First post to any kind of Github project, so please excuse any mistakes in etiquette.

Not _really_ a bug, but for really long tracks the tmpfs can fill up and cause Raspotify to crash.  I recently updated my installation to 0.42.6 and found that Raspotify would crash when playing a 9 hour track (I play white noise to help my infant daughter sleep).

Adding TMPDIR=/tmp to the conf file fixes the problem (revert temp file to main SD card).  Making a note here in case someone else has the same use case.

**Reply by JasonLG1979** [reply]
TMPDIR=/tmp in the config is the default behavior. When you updated last you were told by apt that the config had changed and given the option to keep your config (default), use the new version of the config or open the file to see what changed.

**Reply by JasonLG1979** [reply]
I play it pretty fast and loose, I don't follow Debian version guidelines or any versioning scheme really other than the the 1st digits correspond to the librespot version. So for example 0.42.x uses the librespot 0.4.2 binary. The x in Raspotify 0.42.x is just incremented. So for example if librespot were to come out with 0.4.3 tomorrow my version number would start over and be 0.43.0.

Raspotify otherwise is a rolling release that doesn't guarantee any stability at all as far as the config, systemd unit file or build options. The only guarantee that is made is that it's compatible with Debian Stable.

So if/when apt tells you that the config has changed it's to your advantage to have a look.

**Reply by JasonLG1979** [reply]
But otherwise you're fine no breach of etiquette. I also didn't know that Spotify had 9 hour white noise tracks? Ofc I never really looked for anything like that. You learn something new everyday. That's why I didn't make it the default behavior.

**Reply by dowaichi** [reply]
Thanks, I see the changes to the default config.  Thought I would leave a clue to check the config in case someone else has the same problem.  Error: Os { code: 28, kind: StorageFull, message: "No space left on device" }.

It's funny how temporary solutions become permanent ones.  I started playing the 9 hour white noise track one night when my daughter wouldn't sleep.  I should create a local file so I can play the track offline.

**Reply by JasonLG1979** [reply]
You don't need a local file, you can generate pink noise indefinitely with `speaker-test -c1 -l0 > /dev/null 2>&1`.

**Reply by JasonLG1979** [reply]
You could also go so far as to create an alias of that command called `pink-noise` so you don't have to remember it. That way all you'd have to do is ssh into your device and type `pink-noise` into the terminal.

**Reply by dowaichi** [reply]
Thanks for the awesome tip!  I didn't realize there was something that did this.

**Reply by JasonLG1979** [reply]
Well it's really for testing speaker outputs but the test tone by default is pink noise, `-c1` makes it mono so it doesn't cycle though the speakers, `-l0` makes it loop infinitely and `> /dev/null 2>&1` redirects the terminal output to `/dev/null` otherwise it will fill your terminal with text output. So it's kinda a hack but it works.

**Comment by giggywithit** [comment]
raspi-config
performance options
overlay file system

when you get your system all set up the way you want it, then enable the overlay file system
cheers!

**Reply by JasonLG1979** [reply]
Not a good idea. Raspotify is not designed to be used *that* much like apart of an appliance I update the package very regularly.

**Reply by giggywithit** [reply]
The unit I put together for my mother has been operating for over a year no issues using overlay fs

**Reply by JasonLG1979** [reply]
> The unit I put together for my mother has been operating for over a year no issues using overlay fs

As long as there are no issues and you and her are happy with how it function I guess?
